llvm/test/SYCLLowerIR/PropagateAspectUsagePass/composite_types3.ll

```diff
@@ -0,0 +1,103 @@
+; RUN: opt --PropagateAspectUsage < %s -S | FileCheck %s
+;
+; Test checks structures with pointers.
+
+; By default only %D is optional.
+; %A and %B are optional because they contain %D.
+; %C contains a pointer on %A. Because of that there is a path
+; to %D which means that %C is optional as well.
```

Comment:
```suggestion
; In this test, the only type which is explicitly marked as optional is %D
; Various test cases check if the pass is able to deal with various
; type hierarchies and correctly propagate aspects.
;
; %A and %B are optional because they contain %D.
; %C contains a pointer on %A. Because of that there is a path
; to %D which means that %C is optional as well.
```